File: architecture/serialization-formats/java/src/test/java/tech/khorost/serialization/ExchangeTest.java
package tech.khorost.serialization;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Задача 8, ось перекрёстного чтения: три ловушки обмена через файлы —
 * перепутанное имя, недописанный/испорченный файл, каждый файл несёт
 * хеш и координаты, приём проверяет, что взял именно то, что ожидал.
 * Зеркало Go-теста internal/exchange/exchange_test.go — оба обязаны
 * ловить одни и те же три случая одним и тем же способом (имя файла и
 * форма конверта совпадают побайтово между реализациями).
 */
class ExchangeTest {

    @Test
    void writeThenReadCrossRoundTrips(@TempDir Path dir) {
        byte[] want = {(byte) 0xDE, (byte) 0xAD, (byte) 0xBE, (byte) 0xEF, 0x00, 0x01};
        Exchange.writeCross(dir, "java", "avro", "rename", "newer_reader", 2, want);
        byte[] got = Exchange.readCross(dir, "java", "avro", "rename", "newer_reader", 2);
        assertArrayEquals(want, got);
    }

    @Test
    void readCrossMissingFileRefuses(@TempDir Path dir) {
        assertThrows(Failures.ProbeRefusal.class,
                () -> Exchange.readCross(dir, "java", "avro", "rename", "newer_reader", 0));
    }

    /** «Перепутанное имя»: координаты внутри файла не совпадают с запрошенными. */
    @Test
    void readCrossCoordinateMismatchRefuses(@TempDir Path dir) throws IOException {
        Exchange.writeCross(dir, "java", "avro", "rename", "newer_reader", 0, "payload".getBytes(StandardCharsets.UTF_8));
        Path path = Exchange.crossFileName(dir, "java", "avro", "rename", "newer_reader", 0);
        String raw = Files.readString(path, StandardCharsets.UTF_8);
        String tampered = raw.replace("\"change\":\"rename\"", "\"change\":\"remove\"");
        assertNotEquals(raw, tampered, "подмена не сработала — формат конверта изменился, обнови тест");
        Files.writeString(path, tampered, StandardCharsets.UTF_8);

        assertThrows(Failures.ProbeRefusal.class,
                () -> Exchange.readCross(dir, "java", "avro", "rename", "newer_reader", 0));
    }

    /** Недописанный/испорченный файл: дайджест не совпадает с содержимым. */
    @Test
    void readCrossHashMismatchRefuses(@TempDir Path dir) throws IOException {
        Exchange.writeCross(dir, "go", "protobuf", "unknown_field", "newer_writer", 4,
                "hello world".getBytes(StandardCharsets.UTF_8));
        Path path = Exchange.crossFileName(dir, "go", "protobuf", "unknown_field", "newer_writer", 4);
        String raw = Files.readString(path, StandardCharsets.UTF_8);
        String tampered = raw.replace("\"bytes_b64\":\"", "\"bytes_b64\":\"AA");
        Files.writeString(path, tampered, StandardCharsets.UTF_8);

        assertThrows(Failures.ProbeRefusal.class,
                () -> Exchange.readCross(dir, "go", "protobuf", "unknown_field", "newer_writer", 4));
    }

    @Test
    void writeCrossLeavesNoTempFile(@TempDir Path dir) throws IOException {
        Exchange.writeCross(dir, "java", "json-schema", "add_default", "same", 1,
                "x".getBytes(StandardCharsets.UTF_8));
        try (Stream<Path> entries = Files.list(dir)) {
            assertTrue(entries.noneMatch(p -> p.getFileName().toString().endsWith(".tmp")),
                    "остался временный файл");
        }
    }
}

File: architecture/serialization-formats/go/internal/exchange/exchange_test.go
package exchange

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// Падающий тест (Задача 8, шаг 1): пока пакета exchange нет вовсе, эти
// проверки не компилируются. После реализации они защищают ровно три
// ловушки, названные решением контроллера: перепутанные координаты,
// испорченное/недописанное содержимое и след прошлого прогона —
// каждый файл обмена обязан нести хеш и координаты, а приём — проверять,
// что взял именно то, что ожидал.

func TestWriteThenReadCrossRoundTrips(t *testing.T) {
	dir := t.TempDir()
	want := []byte{0xDE, 0xAD, 0xBE, 0xEF, 0x00, 0x01}

	if err := WriteCross(dir, "go", "avro", "rename", "newer_reader", 2, want); err != nil {
		t.Fatalf("WriteCross: %v", err)
	}
	got, err := ReadCross(dir, "go", "avro", "rename", "newer_reader", 2)
	if err != nil {
		t.Fatalf("ReadCross: %v", err)
	}
	if string(got) != string(want) {
		t.Fatalf("байты разошлись: записали %x, прочитали %x", want, got)
	}
}

func TestReadCrossMissingFile(t *testing.T) {
	dir := t.TempDir()
	if _, err := ReadCross(dir, "go", "avro", "rename", "newer_reader", 0); err == nil {
		t.Fatal("ожидалась ошибка: файла обмена нет вовсе")
	}
}

// TestReadCrossCoordinateMismatch — «перепутанное имя»: файл лежит по
// имени одних координат, а внутри записаны другие. Хеш при этом честный
// (посчитан по байтам, которые реально записаны), поэтому только сверка
// координат ловит подмену.
func TestReadCrossCoordinateMismatch(t *testing.T) {
	dir := t.TempDir()
	if err := WriteCross(dir, "go", "avro", "rename", "newer_reader", 0, []byte("payload")); err != nil {
		t.Fatalf("WriteCross: %v", err)
	}
	// Портим содержимое НЕ трогая имя файла: подменяем change внутри
	// конверта на другое значение, оставляя байты и их хеш согласованными
	// друг с другом, но не с ожидаемыми координатами чтения.
	path := CrossFileName(dir, "go", "avro", "rename", "newer_reader", 0)
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("чтение файла обмена: %v", err)
	}
	tampered := strings.Replace(string(raw), `"change":"rename"`, `"change":"remove"`, 1)
	if tampered == string(raw) {
		t.Fatal("подмена не сработала — формат конверта изменился, обнови тест")
	}
	if err := os.WriteFile(path, []byte(tampered), 0o644); err != nil {
		t.Fatalf("запись подменённого файла: %v", err)
	}

	if _, err := ReadCross(dir, "go", "avro", "rename", "newer_reader", 0); err == nil {
		t.Fatal("ожидался отказ: координаты внутри конверта не совпадают с запрошенными")
	}
}

// TestReadCrossHashMismatch — «недописанный файл» / остаток порчи:
// содержимое не совпадает с зафиксированным при записи дайджестом.
func TestReadCrossHashMismatch(t *testing.T) {
	dir := t.TempDir()
	if err := WriteCross(dir, "java", "protobuf", "unknown_field", "newer_writer", 4, []byte("hello world")); err != nil {
		t.Fatalf("WriteCross: %v", err)
	}
	path := CrossFileName(dir, "java", "protobuf", "unknown_field", "newer_writer", 4)
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("чтение файла обмена: %v", err)
	}
	// Подменяем полезную нагрузку (base64), не трогая записанный дайджест —
	// имитация обрыва записи или порчи посередине.
	tampered := strings.Replace(string(raw), `"bytes_b64":"`, `"bytes_b64":"AA`, 1)
	if err := os.WriteFile(path, []byte(tampered), 0o644); err != nil {
		t.Fatalf("запись подменённого файла: %v", err)
	}

	if _, err := ReadCross(dir, "java", "protobuf", "unknown_field", "newer_writer", 4); err == nil {
		t.Fatal("ожидался отказ: дайджест не совпадает с содержимым")
	}
}

// TestWriteCrossLeavesNoTempFile — запись атомарна: временный файл не
// должен остаться в каталоге обмена после успешной записи.
func TestWriteCrossLeavesNoTempFile(t *testing.T) {
	dir := t.TempDir()
	if err := WriteCross(dir, "go", "json-schema", "add_default", "same", 1, []byte("x")); err != nil {
		t.Fatalf("WriteCross: %v", err)
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatalf("чтение каталога: %v", err)
	}
	for _, e := range entries {
		if filepath.Ext(e.Name()) == ".tmp" {
			t.Fatalf("остался временный файл: %s", e.Name())
		}
	}
}
